docs/proposals/device-crd.md

```diff
@@ -570,6 +570,24 @@ status:
 
 ### Validation
 [Open API v3 Schema based validation](https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions/#validation) can be used to guard against bad requests.
+Invalid values for fields ( example string value for a boolean field etc) can be validated using this.
+In some cases , we also need custom validations (e.g create a device instance which refers to a non -existent device model ) .
+[Validation admission web hooks](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#validatingadmissionwebhook) can be used to implement such custom validation rules.
+
+Here is a list of validations we need to support :
+
+#### Device Model Validations
+- Don't allow model creation if any `Required` fields are missing ( like property name , type, access mode etc.)
+- Property type can be only string or int64. Other types are not supported.
+- The Property access mode for a property can be one of `ReadWrite` or `ReadOnly`. Other access modes are not supported.
+- Property visitors are currently supported for bluetooth, modbus and opcua protocols only. Other protocol visitors are not supported.
+- The BluetoothArithmeticOperationType for Bluetooth read converter can have values one of `[Add, Subtract, Multiply, Divide]`
+- The ModbusRegisterType for Modbus visitor config can have values one of `[CoilRegister, DiscreteInputRegister, InputRegister, HoldingRegister]`. Other register types are not supported.
+
+#### Device Instance Validations
+- Don't allow device instance creation if any `Required` fields are missing ( like devicemodel ref , protocol config, twin property value, twin property name etc.)
+- Don't allow device instance creation if it refers to a device model reference which doesn't exist.
+- Don't allow protocol configs other than opcua, modbus and bluetooth.
```

Comment:
Maybe we can support missing protocol config for wired devices ?

Reply:
currently in the api types , the protocol config is required , but we can see if we can make it optional. updated the doc.

Reply:
https://github.com/kubeedge/examples/blob/master/led-raspberrypi/sample-crds/led-light-device-instance.yaml

I guess this instance works with current types without protocol

Reply:
cc @sujithsimon22

Reply:
@rohitsardesai83 What @sids-b mentioned is true, in the  current example of LED light we do not specify any protocol and the device CRD feature for it seems to be working fine

Reply:
yes what I meant to say was that the api types has a comment saying that is Required, but somehow the openapi schema validation doesn't seem to kick in and so its working. So we can modify the comment in the api types to make it optional. The validation point has already been modified accordingly.

Reply:
I think it might be a good idea for us to have a protocol for wired devices, maybe all types of wired devices can come under such a protocol, but for such a protocol i'm not sure if we need to provide  protocol config or if we do support protocol configs it should be very generic. thoughts?

Reply:
This is what the validation says now : 

- Protocol configs are optional , but if provided , they can be one of `[opcua, modbus and bluetooth]`. Other protocol configs are not supported.

Reply:
I guess that makes sense, in the sense that even if we do support wired protocol they would not require a protocol config, in which case the validation would remain unaffected